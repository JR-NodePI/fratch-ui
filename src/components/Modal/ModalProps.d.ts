import { ReactNode } from 'react';

import { ModalCloseTypes, ModalTypes } from './ModalConstants';

export type ModalType = (typeof ModalTypes)[keyof typeof ModalTypes];

export type ModalCloseType =
  (typeof ModalCloseTypes)[keyof typeof ModalCloseTypes];

export type ModalProps = {
  acceptButtonLabel?: string;
  cancelButtonLabel?: string;
  children: ReactNode;
  onClose?: (type: ModalCloseType) => void;
  onOpen?: () => void;
  title?: ReactNode;
  type?: ModalTypes;
  visible?: boolean;
};

export type ShowModalProps = Omit<
  ModalProps,
  'children' | 'type' | 'visible'
> & { content: ReactNode };

export type ModalContextProps = {
  showModalAccept: (props: ShowModalProps) => void;
  showModalConfirm: (props: ShowModalProps) => void;
  showModalInfo: (props: ShowModalProps) => void;
};
