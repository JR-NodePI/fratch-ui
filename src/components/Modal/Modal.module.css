.modal_overflow {
  position: fixed;
  z-index: 0;
  top: 0;
  left: 0;
  background-color: rgba(0, 0, 0, 0.3);
  width: 100vw;
  height: 100vh;
  display: flex;
  justify-content: space-between;
  align-items: center;
  overflow: auto;
  flex-direction: column;
  pointer-events: all;
  transition: opacity var(--modal-timeout-to-close) ease-in-out;
  z-index: 9999999;
  opacity: 0;
}
.modal_overflow.open {
  opacity: 1;
}
.modal_overflow.close {
  opacity: 0;
  pointer-events: none;
}

.modal_overflow::after,
.modal_overflow::before {
  content: ' ';
  top: 0;
  left: 0;
  width: 100%;
  min-height: 36px;
}

.modal {
  display: flex;
  flex-direction: column;
  background-color: var(--ft-color-background);
  border-radius: var(--ft-button-default-border-radius);
  box-shadow: var(--ft-box-shadow-2);
  max-width: var(--modal-max-width);
  width: 100%;
  transition:
    opacity var(--modal-timeout-to-open) ease-in-out,
    transform var(--modal-timeout-to-open) ease-in-out;
  opacity: 0;
  transform: translateY(-32px);
  outline: solid 1px var(--ft-modal-border-color);
}
.modal_overflow.open .modal {
  opacity: 1;
  transform: translateY(0);
}
.modal_overflow.close .modal {
  opacity: 0;
}
.modal.confirm {
  max-width: var(--modal-confirm-max-width);
}
.modal .header {
  margin: 0;
  border-start-start-radius: var(--ft-button-default-border-radius);
  border-start-end-radius: var(--ft-button-default-border-radius);
  display: flex;
  justify-content: space-between;
  min-height: 42px;
  color: var(--ft-color-lightest);
}
.modal.info .header,
.modal.accept .header {
  background-color: var(--ft-modal-header-bg-color);
}
.modal.info .button_closer svg,
.modal.accept .button_closer svg {
  color: var(--ft-primary-default-color);
}
.modal.confirm .header {
  background-color: var(--ft-modal-header-confirm-bg-color);
}
.modal .header .title {
  margin: 0;
  padding: 12px 24px;
}
.modal .content {
  padding: 16px 42px 16px 24px;
}
.modal .footer {
  gap: 16px;
  display: flex;
  padding: 0 16px 16px 16px;
  justify-content: flex-end;
}
.modal .footer .button {
  min-width: 100px;
}
