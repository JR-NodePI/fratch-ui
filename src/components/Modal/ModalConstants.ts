export const ModalTypes = {
  ACCEPT: 'accept',
  CONFIRM: 'confirm',
  INFO: 'info',
} as const;

export const ModalCloseTypes = {
  ACCEPT: 'accept',
  CANCEL: 'cancel',
  CLOSE: 'close',
} as const;

export const MODAL_TIMEOUT_TO_CLOSE = 500;
export const MODAL_TIMEOUT_TO_OPEN = 200;
export const MODAL_MAX_WIDTH = 600;
export const MODAL_CONFIRM_MAX_WIDTH = 460;
