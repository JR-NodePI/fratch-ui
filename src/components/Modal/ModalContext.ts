import React from 'react';

import type { ModalContextProps } from './ModalProps';

const ModalContext = React.createContext<ModalContextProps>({
  showModalAccept: () => {},
  showModalConfirm: () => {},
  showModalInfo: () => {},
});
ModalContext.displayName = 'ModalContext';

export default ModalContext;
