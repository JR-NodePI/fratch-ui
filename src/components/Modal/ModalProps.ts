import { ReactNode } from 'react';

export const ModalTypes = {
  ACCEPT: 'accept' as const,
  CONFIRM: 'confirm' as const,
  INFO: 'info' as const,
};

export const ModalCloseTypes = {
  ACCEPT: 'accept' as const,
  CANCEL: 'cancel' as const,
  CLOSE: 'close' as const,
};

export type ModalCloseType = (typeof ModalCloseTypes)[keyof typeof ModalCloseTypes];

export type ModalProps = {
  acceptButtonLabel?: string;
  cancelButtonLabel?: string;
  children: ReactNode;
  onClose?: (type: ModalCloseType) => void;
  onOpen?: () => void;
  title?: ReactNode;
  type?: (typeof ModalTypes)[keyof typeof ModalTypes];
  visible?: boolean;
};
