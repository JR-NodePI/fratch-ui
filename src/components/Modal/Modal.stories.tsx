import type { Meta, StoryObj } from '@storybook/react';

import Modal from './Modal';

const meta = {
  title: 'Example/Modal',
  component: Modal,
  argTypes: {
    onClose: { action: 'onClose', table: { disable: true } },
    onOpen: { action: 'onOpen' },
  },
  tags: ['autodocs'],
} satisfies Meta<typeof Modal>;

export default meta;

type Story = StoryObj<typeof Modal>;

export const Docs: Story = {
  args: {
    title: 'This is the title',
    children: 'This is the content',
    visible: true,
  },
};
