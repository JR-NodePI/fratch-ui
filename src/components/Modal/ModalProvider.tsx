import { ReactNode, useCallback, useMemo, useState } from 'react';

import Modal from './Modal';
import { ModalTypes } from './ModalConstants';
import ModalContext from './ModalContext';
import {
  ModalCloseType,
  ModalContextProps,
  ModalProps,
  ModalType,
  ShowModalProps,
} from './ModalProps';

const defaultModalProps = {
  children: <></>,
  title: '',
  visible: false,
};

function ModalProvider({ children }: { children: ReactNode }): JSX.Element {
  const [modalProps, setModalProps] = useState<ModalProps>(defaultModalProps);

  const handleModalOpen = useCallback(
    (props: ShowModalProps, type: ModalType) => {
      setModalProps(state => {
        if (state.visible) {
          return state;
        }

        const handleClose = (type: ModalCloseType): void => {
          setModalProps(defaultModalProps);
          props.onClose?.(type);
        };

        return {
          ...props,
          children: props.content,
          onClose: handleClose,
          type,
          visible: !state.visible,
        };
      });
    },
    []
  );

  const value = useMemo<ModalContextProps>(
    () => ({
      showModalAccept: (props: ShowModalProps): void => {
        handleModalOpen(props, ModalTypes.ACCEPT);
      },
      showModalConfirm: (props): void => {
        handleModalOpen(props, ModalTypes.CONFIRM);
      },
      showModalInfo: (props): void => {
        handleModalOpen(props, ModalTypes.INFO);
      },
    }),
    [handleModalOpen]
  );

  return (
    <ModalContext.Provider value={value}>
      {children}
      <Modal {...modalProps} />
    </ModalContext.Provider>
  );
}

export default ModalProvider;
