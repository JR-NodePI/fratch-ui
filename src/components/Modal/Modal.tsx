import { createPortal } from 'react-dom';
import { useCallback, useEffect, useState } from 'react';
import { ModalCloseTypes, type ModalProps, ModalTypes, type ModalCloseType } from './ModalProps';
import Button from '../Button/Button';
import { v4 as uuid } from 'uuid';
import { c } from '../../helpers/classNameHelpers';

import styles from './Modal.module.css';
import { isAscendantEvenTargetByID } from '../../helpers/htmlSelectorsHelpers';

function Modal({
  acceptButtonLabel,
  cancelButtonLabel,
  children,
  onClose,
  onOpen,
  title,
  type = ModalTypes.INFO,
  visible,
}: ModalProps): JSX.Element {
  const [id] = useState<string>(uuid());
  const [cssClassStatus, setCssClassStatus] = useState<string>('');
  const [mounted, setMounted] = useState<boolean>(false);
  const [isTransitioning, setIsTransitioning] = useState<boolean>(false);

  const close = useCallback((): void => {
    setIsTransitioning(true);
    setCssClassStatus(styles.close);

    setTimeout(() => {
      setCssClassStatus('');
      setMounted(false);
      setIsTransitioning(false);
    }, 500);
  }, []);

  const open = useCallback((): void => {
    setIsTransitioning(true);
    setMounted(true);

    setTimeout(() => {
      onOpen?.();
      setCssClassStatus(styles.open);
      setIsTransitioning(false);
    }, 100);
  }, [onOpen]);

  useEffect(() => {
    visible ? open() : close();
  }, [close, open, visible]);

  const handleCloseTypes = (type: ModalCloseType = ModalCloseTypes.CLOSE): void => {
    if (isTransitioning) {
      return;
    }
    onClose?.(type);
    close();
  };

  const handleAccept = (): void => {
    handleCloseTypes(ModalCloseTypes.ACCEPT);
  };
  const handleCancel = (): void => {
    handleCloseTypes(ModalCloseTypes.CANCEL);
  };
  const handleOverflowClose = (event: React.MouseEvent): void => {
    if (!isAscendantEvenTargetByID(event.nativeEvent, id)) {
      handleCloseTypes(ModalCloseTypes.CANCEL);
    }
  };
  const handleOverflow = (): void => {
    handleCloseTypes(ModalCloseTypes.CANCEL);
  };

  const hasHeader = title || type === ModalTypes.INFO;
  const hasCloser = type !== ModalTypes.CONFIRM;
  const hasFooter = type !== ModalTypes.INFO;

  if (!mounted) {
    return <></>;
  }

  return createPortal(
    <div
      className={c(styles.modal_overflow, cssClassStatus)}
      onClick={hasCloser ? handleOverflowClose : undefined}
    >
      <section id={id} className={c(styles.modal, styles[type])}>
        {hasHeader && (
          <header className={c(styles.header)}>
            <h5 className={c(styles.title)}>{title}</h5>
            {hasCloser && (
              <button className={c(styles.closer)} onClick={handleOverflow}>
                Ã—
              </button>
            )}
          </header>
        )}
        <div className={c(styles.content)}>{children}</div>{' '}
        {hasFooter && (
          <footer className={c(styles.footer)}>
            {type === ModalTypes.CONFIRM && (
              <Button
                onClick={handleCancel}
                className={c(styles.button)}
                label={cancelButtonLabel || 'Cancel'}
                size="small"
              />
            )}

            <Button
              onClick={handleAccept}
              className={c(styles.button)}
              type="primary"
              label={acceptButtonLabel || (type === ModalTypes.CONFIRM ? 'OK' : 'Accept')}
              size="small"
            />
          </footer>
        )}
      </section>
    </div>,
    document.body
  );
}

export default Modal;
