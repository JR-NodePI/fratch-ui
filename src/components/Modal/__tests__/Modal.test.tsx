import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import debounce from 'lodash/debounce';
import { describe, expect, it, vi } from 'vitest';

import Modal from '../Modal';
import { ModalTypes } from '../ModalConstants';
import { type ModalProps } from '../ModalProps';

vi.mock('lodash/debounce', async requireActual => {
  const module = (await requireActual()) as { default: typeof debounce };
  return { default: vi.fn(fn => module.default(fn, 10)) };
});

describe('Modal', () => {
  const modalId = 'crypto-mock-random-UUID-1234';
  const initialProps: Omit<ModalProps, 'children'> = {};

  const getComponent = (props = initialProps): JSX.Element => (
    <Modal {...props}>
      <p>Modal contents</p>
    </Modal>
  );

  const setup = (props = initialProps): ReturnType<typeof render> =>
    render(getComponent(props));

  beforeEach(() => {
    vi.spyOn(crypto, 'randomUUID').mockReturnValue(modalId);
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('should render properly', () => {
    const { baseElement } = setup();
    expect(baseElement).toMatchSnapshot();
  });

  it('should render as visible', () => {
    const { baseElement } = setup({
      visible: true,
      title: 'Modal title',
    });
    expect(baseElement).toMatchSnapshot();
  });

  it('should render for type ACCEPT', () => {
    const { baseElement } = setup({
      visible: true,
      type: ModalTypes.ACCEPT,
    });
    expect(baseElement).toMatchSnapshot();
  });

  it('should render for type CONFIRM with custom button labels', () => {
    const { baseElement } = setup({
      visible: true,
      type: ModalTypes.CONFIRM,
      acceptButtonLabel: 'mock-custom-label-OK',
      cancelButtonLabel: 'mock-custom-label-CANCEL',
    });
    expect(baseElement).toMatchSnapshot();
  });

  it('should perform onOpen on rerender as visible', async () => {
    const onOpen = vi.fn();
    const { rerender } = setup();

    await rerender(getComponent({ visible: true, onOpen }));

    expect(onOpen).toHaveBeenCalledTimes(1);
    expect(onOpen).toHaveBeenCalledWith();

    expect(debounce).toHaveBeenCalledTimes(1);
    expect(debounce).toHaveBeenNthCalledWith(1, expect.any(Function), 100);
  });

  it('should perform onClose with accept for type CONFIRM', async () => {
    const onClose = vi.fn();
    setup({
      visible: true,
      onClose,
      type: ModalTypes.CONFIRM,
    });

    const closeButton = screen.getByRole('button', { name: /OK/i });
    await userEvent.click(closeButton);

    await waitFor(() => {
      expect(onClose).toHaveBeenCalledTimes(1);
      expect(onClose).toHaveBeenCalledWith('accept');
    });

    expect(debounce).toHaveBeenCalledTimes(2);
    expect(debounce).toHaveBeenNthCalledWith(1, expect.any(Function), 100);
    expect(debounce).toHaveBeenNthCalledWith(2, expect.any(Function), 500);
  });

  it('should perform onClose with cancel for type CONFIRM', async () => {
    const onClose = vi.fn();
    setup({
      visible: true,
      onClose,
      type: ModalTypes.CONFIRM,
    });

    const closeButton = screen.getByRole('button', { name: /Cancel/i });
    await userEvent.click(closeButton);

    await waitFor(() => {
      expect(onClose).toHaveBeenCalledTimes(1);
      expect(onClose).toHaveBeenCalledWith('cancel');
    });

    expect(debounce).toHaveBeenCalledTimes(2);
    expect(debounce).toHaveBeenNthCalledWith(1, expect.any(Function), 100);
    expect(debounce).toHaveBeenNthCalledWith(2, expect.any(Function), 500);
  });

  it('should perform onClose from closer button', async () => {
    const onClose = vi.fn();
    const { baseElement } = setup({
      visible: true,
      onClose,
    });

    const closeButton = screen.getByRole('button', { name: /Close/i });
    await userEvent.click(closeButton);

    await waitFor(() => {
      expect(onClose).toHaveBeenCalledTimes(1);
      expect(onClose).toHaveBeenCalledWith('close');
    });

    expect(debounce).toHaveBeenCalledTimes(2);
    expect(debounce).toHaveBeenNthCalledWith(1, expect.any(Function), 100);
    expect(debounce).toHaveBeenNthCalledWith(2, expect.any(Function), 500);

    expect(baseElement).toMatchSnapshot();
  });

  it('should perform onClose from overflow', async () => {
    const onClose = vi.fn();
    const { baseElement } = setup({
      visible: true,
      onClose,
    });

    const overflow = screen.getByLabelText('Close modal');
    await userEvent.click(overflow);

    await waitFor(() => {
      expect(onClose).toHaveBeenCalledTimes(1);
      expect(onClose).toHaveBeenCalledWith('close');
    });

    expect(debounce).toHaveBeenCalledTimes(2);
    expect(debounce).toHaveBeenNthCalledWith(1, expect.any(Function), 100);
    expect(debounce).toHaveBeenNthCalledWith(2, expect.any(Function), 500);

    expect(baseElement).toMatchSnapshot();
  });
});
