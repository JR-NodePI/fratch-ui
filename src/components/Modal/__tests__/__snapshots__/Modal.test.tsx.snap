// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal > should perform onClose from closer button 1`] = `
<body>
  <div />
</body>
`;

exports[`Modal > should perform onClose from overflow 1`] = `
<body>
  <div />
</body>
`;

exports[`Modal > should render as visible 1`] = `
<body>
  <div />
  <div
    aria-label="Close modal"
    class="_modal_overflow_f57fcc"
    style="--modal-timeout-to-close: 500ms; --modal-timeout-to-open: 200ms; --modal-max-width: 600px; --modal-confirm-max-width: 460px;"
  >
    <section
      class="_modal_f57fcc _info_f57fcc"
      id="crypto-mock-random-UUID-1234"
    >
      <header
        class="_header_f57fcc"
      >
        <h5
          class="_title_f57fcc"
        >
          Modal title
        </h5>
      </header>
      <button
        class="_button_closer_8d9a3f _button_closer_f57fcc"
        title="Close"
      >
        <svg
          aria-hidden="true"
          class="_icon_2be8fc _primary_2be8fc _icon_8d9a3f"
          fill="none"
          viewBox="-1 -1 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <i>
          Close
        </i>
      </button>
      <div
        class="_content_f57fcc"
      >
        <p>
          Modal contents
        </p>
      </div>
    </section>
  </div>
</body>
`;

exports[`Modal > should render for type ACCEPT 1`] = `
<body>
  <div />
  <div
    aria-label="Close modal"
    class="_modal_overflow_f57fcc"
    style="--modal-timeout-to-close: 500ms; --modal-timeout-to-open: 200ms; --modal-max-width: 600px; --modal-confirm-max-width: 460px;"
  >
    <section
      class="_modal_f57fcc _accept_f57fcc"
      id="crypto-mock-random-UUID-1234"
    >
      <button
        class="_button_closer_8d9a3f _button_closer_f57fcc"
        title="Close"
      >
        <svg
          aria-hidden="true"
          class="_icon_2be8fc _primary_2be8fc _icon_8d9a3f"
          fill="none"
          viewBox="-1 -1 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <i>
          Close
        </i>
      </button>
      <div
        class="_content_f57fcc"
      >
        <p>
          Modal contents
        </p>
      </div>
      <footer
        class="_footer_f57fcc"
      >
        <button
          class="_button_2a1efe _primary_2a1efe _small_2a1efe _button_f57fcc"
        >
          <span>
            Accept
          </span>
        </button>
      </footer>
    </section>
  </div>
</body>
`;

exports[`Modal > should render for type CONFIRM with custom button labels 1`] = `
<body>
  <div />
  <div
    aria-label=""
    class="_modal_overflow_f57fcc"
    style="--modal-timeout-to-close: 500ms; --modal-timeout-to-open: 200ms; --modal-max-width: 600px; --modal-confirm-max-width: 460px;"
  >
    <section
      class="_modal_f57fcc _confirm_f57fcc"
      id="crypto-mock-random-UUID-1234"
    >
      <div
        class="_content_f57fcc"
      >
        <p>
          Modal contents
        </p>
      </div>
      <footer
        class="_footer_f57fcc"
      >
        <button
          class="_button_2a1efe _default_2a1efe _small_2a1efe _button_f57fcc"
        >
          <span>
            mock-custom-label-CANCEL
          </span>
        </button>
        <button
          class="_button_2a1efe _tertiary_2a1efe _small_2a1efe _button_f57fcc"
        >
          <span>
            mock-custom-label-OK
          </span>
        </button>
      </footer>
    </section>
  </div>
</body>
`;

exports[`Modal > should render properly 1`] = `
<body>
  <div />
</body>
`;
